legacy-backend-rust/src/main.rs

```diff
@@ -4,47 +4,59 @@ use async_std::{
     task,
 };
 use rand::{rngs::SmallRng, Rng, SeedableRng};
-use serde_json::Value;
-use std::time::Duration;
-use tide::{http::mime, Request, Response};
+use rand_distr::Normal;
+use std::{collections::HashMap, time::Duration};
+use tide::{http::mime, Body, Error, Request, Response, StatusCode};
+
+const DELAY_SECS_MEAN: f64 = 2.0;
+const DELAY_STD_DEV: f64 = 0.3;
 
 #[derive(Clone)]
 struct State {
-    data: Vec<Value>,
+    data: HashMap<&'static str, Vec<u8>>,
     rng: Arc<Mutex<SmallRng>>,
+    normal: Normal<f64>,
 }
 
 async fn handle_request(req: Request<State>) -> tide::Result {
     let state = req.state();
 
-    let (delay, index) = {
+    let data_type: String = req.param("data_type")?;
+    let data = state
+        .data
+        .get(data_type.as_str())
+        .ok_or_else(|| Error::from_str(StatusCode::NotFound, "no data"))?;
+
+    let delay = {
         let arc = state.rng.clone();
         let mut rng = arc.lock().await;
-        (
-            rng.gen_range(1000, 3000),
-            rng.gen_range(0, state.data.len() - 1),
-        )
+        rng.sample::<f64, _>(state.normal) as u64
     };
     task::sleep(Duration::from_millis(delay)).await;
 
-    let res = Response::builder(200)
-        .body(state.data[index].to_string())
-        .content_type(mime::JSON)
-        .build();
+    let mut res = Response::new(200);
+    res.set_body(Body::from_bytes(data.clone()));
+    res.set_content_type(mime::JSON);
     Ok(res)
 }
 #[async_std::main]
 async fn main() -> Result<()> {
     // tide::log::start();
-    let data = include_str!("../data/generated-1.json");
-    let data: Value = serde_json::from_str(data)?;
-    let data = data
-        .as_array()
-        .ok_or_else(|| anyhow::anyhow!("data is not an array"))?
-        .to_vec();
+    let accounts = include_bytes!("../data/accounts.json").to_vec();
+    let cards = include_bytes!("../data/cards.json").to_vec();
+    let customer = include_bytes!("../data/customer.json").to_vec();
+
     let rng = Arc::new(Mutex::new(SmallRng::from_entropy()));
-    let mut app = tide::with_state(State { data, rng });
-    app.at("/").get(handle_request);
-    app.listen("127.0.0.1:8080").await?;
+    let normal = Normal::new(1000.0 * DELAY_SECS_MEAN, 1000.0 * DELAY_STD_DEV).unwrap();
+
+    let mut data = HashMap::new();
+    data.insert("accounts", accounts);
+    data.insert("cards", cards);
+    data.insert("customer", customer);
+
+    let mut app = tide::with_state(State { data, rng, normal });
```

Comment:
Not at all important, but I wonder whether you could avoid the state entirely by capturing the RNG and the relevant data in a closure of the handler. It may also be a tiny bit more performant if each endpoint has its own RNG, avoiding the mutex.

Random thoughts, more to check my own understanding of the options available.

Reply:
Yeah, I thought about doing what you'd done (so easily) in the go version. But have you ever tried returning an async function from another function? Not particularly straightforward.
Incidentally, I think you'd still need a Mutex as there's no guarantee that each request will be handled by the same thread.